docs/notebooks/implicit_diff/maml.md

```diff
@@ -0,0 +1,275 @@
+    in_params_sol, _ = lbfgs.run(in_params, outer_params, data_train)
+    prediction = model.apply(in_params_sol, x_test)
+    loss += jnp.mean((prediction - y_test)**2)
+    task_params.append(in_params_sol)
+  return loss, task_params
+```
+
+```{code-cell} ipython3
+key, subkey = random.split(random.PRNGKey(0), 2)
+
+# initialize inner and outer params
+inner_params = []
+for _ in data_tasks:
+  key, subkey = random.split(key)
+  inner_params.append(model.init(key, dummy_input))
+
+key, subkey = random.split(key)
+outer_params = model.init(key, dummy_input)
+
+gradient_subopt = []
+
+solver = OptaxSolver(opt=optax.adam(1e-3), fun=outer_loss, maxiter=100, has_aux=True)
+state = solver.init_state(outer_params, inner_params, data_tasks[:-1])
+jitted_update = jax.jit(solver.update)
+
+for it in range(solver.maxiter):
+  outer_params, state = jitted_update(outer_params, state, state.aux, data_tasks[:-1])
+  gradient_subopt.append(solver.l2_optimality_error(outer_params, state.aux, data_tasks[:-1]))
+```
+
+```{code-cell} ipython3
+xx = jnp.linspace(MIN_X, MAX_X, 200)
+plt.title(f'Training data and predictive model')
+for task, ((x_train, y_train), test) in enumerate(data_tasks[:-1]):
+  plt.scatter(x_train, y_train, marker='o', s=50, label=f"Training samples for task {task+1}", color=colors[task])
+  prediction = jax.lax.stop_gradient(model.apply(state.aux[task], xx.reshape((-1, 1))))
+  plt.plot(xx, prediction.ravel(), color=colors[task], lw=3, label=f"Prediction of model trained on task {task+1}")
+plt.legend(loc='upper center', fontsize=14, bbox_to_anchor=(0.5, -0.1), frameon=False, ncol=3)
+plt.xlabel('x')
+plt.ylabel('y')
+plt.show()
+
+plt.title('Gradient suboptimality')
+plt.plot(gradient_subopt, lw=3)
+plt.yscale('log')
+plt.ylabel("gradient norm of outer objective")
+plt.xlabel("iterations")
+plt.grid()
+plt.show()
+```
+
+```{code-cell} ipython3
+# now we consider the last task, which we haven't used for training
+train, test = data_tasks[-1]
+
+params, _ = LBFGS(
+    fun=inner_loss, tol=1e-12, maxiter=2000).run(
+    model.init(key, dummy_input), outer_params, train)
+
+params_without_regularization, _ = LBFGS(
+    fun=inner_loss, tol=1e-12, maxiter=2000).run(
+    model.init(key, dummy_input), outer_params, train, regularization=0)
+```
+
+```{code-cell} ipython3
+xx = jnp.linspace(MIN_X, MAX_X, 200)
+prediction = model.apply(params, xx.reshape((-1, 1)))
+prediction_without_regularization = model.apply(params_without_regularization, xx.reshape((-1, 1)))
+
+plt.plot(xx, prediction.ravel(), lw=3, label="with meta-learning")
+plt.plot(xx, prediction_without_regularization.ravel(), lw=3, label="without meta-learning")
+plt.scatter(train[0], train[1], marker='^', s=80, label=f"Training samples", color=colors[task])
+plt.plot(xs, ys, lw=3, label="ground truth")
+plt.ylim((-1, 1))
+plt.xlim(MIN_X, MAX_X)
+plt.legend(loc='upper center', bbox_to_anchor=(0.5, -0.1), frameon=False, ncol=n_tasks)
+plt.grid()
```

Comment:
To convey the difficulty of the task I would add something like `plt.title("Fit with only 3 samples")`

Reply:
nice suggestion! implemented